homeassistant/components/media_player/heos.py

```diff
@@ -0,0 +1,217 @@
+
+    heos = HeosMediaPlayer(hass, host, name, username, password)
+
+    yield from heos.heos.connect(
+        host=host, callback=heos.async_update_ha_state)
+
+    async_add_devices([heos])
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, hass, host, name, username, password):
+        """Initialize."""
+        from aioheos import AioHeos
+        if not host:
+            _LOGGER.info('No host provided, will try to discover...')
+        self._hass = hass
+        _verbose = False
+        self.heos = AioHeos(
+            loop=hass.loop,
+            host=host,
+            username=username,
+            password=password,
+            verbose=_verbose)
+        self._name = name
+        self._state = None
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Retrieve latest state."""
+        self.heos.request_play_state()
+        self.heos.request_mute_state()
+        self.heos.request_volume()
+        self.heos.request_now_playing_media()
+        return True
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def volume_level(self):
+        """Volume level of the device (0..1)."""
+        volume = self.heos.get_volume()
+        return float(volume) / 100
+
+    @property
+    def state(self):
+        """Get state."""
+        self._state = self.heos.get_play_state()
+        if self._state == 'stop':
+            return STATE_OFF
+        if self._state == 'pause':
+            return STATE_PAUSED
+        if self._state == 'play':
+            return STATE_PLAYING
+        return STATE_UNKNOWN
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def media_content_type(self):
+        """Content type of current playing media."""
+        return MEDIA_TYPE_MUSIC
+
+    @property
+    def media_artist(self):
+        """Artist of current playing media."""
+        return self.heos.get_media_artist()
+
+    @property
+    def media_title(self):
+        """Album name of current playing media."""
+        return self.heos.get_media_song()
+
+    @property
+    def media_album_name(self):
+        """Album name of current playing media."""
+        return self.heos.get_media_album()
+
+    @property
+    def media_image_url(self):
+        """Return the image url of current playing media."""
+        return self.heos.get_media_image_url()
+
+    @property
+    def media_content_id(self):
+        """Return the content ID of current playing media."""
+        return self.heos.get_media_id()
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        muted_state = self.heos.get_mute_state()
+        return muted_state == 'on'
+
+    @asyncio.coroutine
+    def async_mute_volume(self, _mute):
+        """Mute volume."""
+        self.heos.toggle_mute()
```

Comment:
This function is not a toggle.  Based on the value of `mute` it either needs to set or unset mute on the underlying device.  The CLI has a command for this: `heos://player/set_mute?pid=player_id&state=on_or_off`